### PR TITLE
fix: split "union" and "interleave"

### LINKED ISSUE #5970: UNION ALL with ORDER BY results are inconsistent
### Describe the bug

The results for a `UNION ALL` that specifies an `ORDER BY` is sometimes sorted incorrectly.

> **Note**
>
> I bisected to the first bad commit being 5fc91cc8fbb56b6d2a32e66f8b327a871f7d15ac via #5661

### To Reproduce

Using `datafusion-cli` (version 22.0) execute;

```sql
WITH
  m0(time, tag0, f64) AS (VALUES (1667181600000000000, 'val00', 10.1), (1667181610000000000, 'val00', 21.2), (1667181620000000000, 'val00', 11.2), (1667181630000000000, 'val00', 19.2), (1667181600000000000, 'val01', 11.3), (1667181600000000000, 'val02', 10.4), (1667181610000000000, 'val00', 18.9)),

  m1(time, tag0, f64) AS (VALUES (1667181600000000000, 'val00', 100.5), (1667181610000000000, 'val00', 200.6), (1667181600000000000, 'val01', 101.7)),

  t AS (
    SELECT 'm0' as "iox::measurement", tag0, 0::timestamp as time, COUNT(f64), SUM(f64), stddev(f64) FROM m0 GROUP BY 1, 2, 3), 

  u AS (
    SELECT 'm1' as "iox::measurement", tag0, 0::timestamp as time, COUNT(f64), SUM(f64), stddev(f64) FROM m1 GROUP BY 1, 2, 3)
SELECT * FROM t
UNION ALL
SELECT * FROM u
ORDER BY 1, 2, 3;
```

### Expected behavior

Output should be sorted on the first 3 columns:

```text
+------------------+-------+---------------------+------------+----------+-------------------+
| iox::measurement | tag0  | time                | COUNT(f64) | SUM(f64) | STDDEV(f64)       |
+------------------+-------+---------------------+------------+----------+-------------------+
| m0               | val00 | 1970-01-01T00:00:00 | 5          | 80.6     | 5.085961069453836 |
| m0               | val01 | 1970-01-01T00:00:00 | 1          | 11.3     |                   |
| m0               | val02 | 1970-01-01T00:00:00 | 1          | 10.4     |                   |
| m1               | val00 | 1970-01-01T00:00:00 | 2          | 301.1    | 70.7813887967734  |
| m1               | val01 | 1970-01-01T00:00:00 | 1          | 101.7    |                   |
+------------------+-------+---------------------+------------+----------+-------------------+
```

### Actual behavior

Output is sometimes inconsistent, noting that the sort order is incorrect:

```text
+------------------+-------+---------------------+------------+----------+-------------------+
| iox::measurement | tag0  | time                | COUNT(f64) | SUM(f64) | STDDEV(f64)       |
+------------------+-------+---------------------+------------+----------+-------------------+
| m0               | val00 | 1970-01-01T00:00:00 | 5          | 80.6     | 5.085961069453836 |
| m0               | val01 | 1970-01-01T00:00:00 | 1          | 11.3     |                   |
| m1               | val00 | 1970-01-01T00:00:00 | 2          | 301.1    | 70.7813887967734  |
| m0               | val02 | 1970-01-01T00:00:00 | 1          | 10.4     |                   |
| m1               | val01 | 1970-01-01T00:00:00 | 1          | 101.7    |                   |
+------------------+-------+---------------------+------------+----------+-------------------+
```

### Additional context

DataFusion versions 18, 19, 20 and 21 work consistently after numerous executions of the same query

#### The following also works on 22

```sql
WITH
  m0(time, tag0, f64) AS (VALUES (1667181600000000000, 'val00', 10.1), (1667181610000000000, 'val00', 21.2), (1667181620000000000, 'val00', 11.2), (1667181630000000000, 'val00', 19.2), (1667181600000000000, 'val01', 11.3), (1667181600000000000, 'val02', 10.4), (1667181610000000000, 'val00', 18.9)),
  m1(time, tag0, f64) AS (VALUES (1667181600000000000, 'val00', 100.5), (1667181610000000000, 'val00', 200.6), (1667181600000000000, 'val01', 101.7)),
  t AS (SELECT 'm0' as "iox::measurement", tag0, 0::timestamp as time, COUNT(f64), SUM(f64), stddev(f64) FROM m0 GROUP BY 1, 2, 3), 
  u AS (SELECT 'm1' as "iox::measurement", tag0, 0::timestamp as time, COUNT(f64), SUM(f64), stddev(f64) FROM m1 GROUP BY 1, 2, 3), 
-- add another CTE
  v AS (
  SELECT * FROM t
  UNION ALL
  SELECT * FROM u)
SELECT * FROM v
ORDER BY 1, 2, 3;
```

### PR DESCRIPTION
**Pull request by crepererum**
# Which issue does this PR close?
Fixes #5970.

This is an alternative to #6009. Closes https://github.com/apache/arrow-datafusion/pull/6009

# Rationale for this change
`UnionExec` shall be a simple node and not perform some optimizations on its own. If repartition / partition concatenation is desired, then there should be an optimizer pass for that. Being a simple node avoids confusion and optimizer bugs like #5970.

# What changes are included in this PR?
Remove the entire "partition aware" logic from `UnionExec`

# Are these changes tested?
1. Existing tests pass
2. Test for #5970 passes.
3. Query mentioned in [his comment](https://github.com/apache/arrow-datafusion/pull/6045#issuecomment-1513390720) is correctly optimized (i.e. does not require top-level repartition)

# Are there any user-facing changes?
Less bugs.

### CODE REVIEW

**Comment by mingmwang** [comment]
Could you please add a UT for the rule itself ?
For example for the case there are multiple partition-aware UnionExecs, finally they are all replaced.

**Comment by crepererum** [comment]
will add a unit test for the optimizer. will also try to add a test for the query listed above. I think this can be fixed by pushing the `Interleave` optimizer rule before `Repartition`.

**Comment by crepererum** [comment]
I have this working locally by changing the optimizer. Instead of having an extra optimizer, this has to be integrated into the "dist enforcement" transform-up system. I'll need to add more tests though and clean up the code, but I think the result is satisfying.

**Comment by mingmwang** [comment]
> I have this working locally by changing the optimizer. Instead of having an extra optimizer, this has to be integrated into the "dist enforcement" transform-up system. I'll need to add more tests though and clean up the code, but I think the result is satisfying.

Yes, this should works, thanks!

File: datafusion/core/src/physical_plan/union.rs [hunk 1]
```diff
@@ -273,36 +230,210 @@ impl ExecutionPlan for UnionExec {
...
-                if partition < input.output_partitioning().partition_count() {
-                    let stream = input.execute(partition, context)?;
-                    debug!("Found a Union partition to execute");
-                    return Ok(Box::pin(ObservedStream::new(stream, baseline_metrics)));
-                } else {
-                    partition -= input.output_partitioning().partition_count();
-                }
+            vec![false; self.inputs().len()]
+        }
+    }
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:400]: I think we can return false here.

File: datafusion/core/src/physical_plan/union.rs [hunk 2]
```diff
@@ -273,36 +230,210 @@ impl ExecutionPlan for UnionExec {
...
         }
+        if input_stream_vec.len() == self.inputs.len() {
+            let stream = Box::pin(CombinedRecordBatchStream::new(
+                self.schema(),
+                input_stream_vec,
+            ));
+            return Ok(Box::pin(ObservedStream::new(stream, baseline_metrics)));
+        }
 
         warn!("Error in Union: Partition {} not found", partition);
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:439]: Please change the warning message.

Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: done

File: datafusion/core/src/physical_plan/union.rs (hunk 1 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: fixed

File: datafusion/core/src/physical_plan/union.rs [hunk 3]
```diff
@@ -340,8 +278,222 @@ impl ExecutionPlan for UnionExec {
...
+    fn unbounded_output(&self, children: &[bool]) -> Result<bool> {
+        Ok(children.iter().any(|x| *x))
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        self.inputs.clone()
+    }
+
+    /// Output of the union is the combination of all output partitions of the inputs
+    fn output_partitioning(&self) -> Partitioning {
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:369]: Please fix this comment also.

File: datafusion/core/src/physical_plan/union.rs [hunk 4]
```diff
@@ -340,8 +278,222 @@ impl ExecutionPlan for UnionExec {
...
+/// +---------+
+/// ```
+#[derive(Debug)]
+pub struct InterleaveExec {
+    /// Input execution plan
+    inputs: Vec<Arc<dyn ExecutionPlan>>,
+    /// Execution metrics
+    metrics: ExecutionPlanMetricsSet,
+    /// Schema of Union
+    schema: SchemaRef,
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:320]: Schema of `InterleaveExec`

File: datafusion/core/src/physical_plan/union.rs [hunk 5]
```diff
@@ -340,8 +278,222 @@ impl ExecutionPlan for UnionExec {
...
+                self.schema(),
+                input_stream_vec,
+            ));
+            return Ok(Box::pin(ObservedStream::new(stream, baseline_metrics)));
+        }
+
+        warn!("Error in InterleaveExec: Partition {} not found", partition);
+
+        Err(crate::error::DataFusionError::Execution(format!(
+            "Partition {partition} not found in Union"
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:420]: "Partition {partition} not found in InterleaveExec"

**Review by mingmwang — APPROVED** [review]

File: datafusion/core/src/physical_plan/union.rs (hunk 3 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: fixed

File: datafusion/core/src/physical_plan/union.rs (hunk 4 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: fixed

File: datafusion/core/src/physical_plan/union.rs (hunk 5 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: fixed

File: datafusion/core/src/physical_plan/union.rs [hunk 6]
```diff
@@ -340,8 +278,222 @@ impl ExecutionPlan for UnionExec {
...
+    /// infinite, returns an error to indicate this.
+    fn unbounded_output(&self, children: &[bool]) -> Result<bool> {
+        Ok(children.iter().any(|x| *x))
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        self.inputs.clone()
+    }
+
+    /// Output of the interleave is the combination of all output partitions of the inputs
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/union.rs:368]: I think it is not the combination of all output partitions of the inputs. It keeps one of the input's output partition.

**Comment by ozankabak** [comment]
Does the new `InterleaveExec` behave similar to `UnionExec` in terms of sort enforcement? There are a couple places in `sort_enforcement.rs` where it makes a call to `is_union` to optimize/enforce sorts efficiently. Do we need to make any changes there?

@mustafasrepo, PTAL

**Comment by mingmwang** [comment]
> InterleaveExec

`InterleaveExec` can not keep the ordering(`maintains_input_order` is false), I think the logic in the rules do not need to update.

File: datafusion/core/src/physical_plan/union.rs (hunk 6 above)
Comment (crepererum) [reply datafusion/core/src/physical_plan/union.rs]: fixed

**Comment by mingmwang** [comment]
I think this PR can be merged to main.

**Comment by alamb** [comment]
Thanks @mingmwang

**Pull request merged**